render/vulkan/texture.c

```diff
@@ -0,0 +1,715 @@
+	img_info.samples = VK_SAMPLE_COUNT_1_BIT;
+	img_info.sharingMode = VK_SHARING_MODE_EXCLUSIVE;
+	img_info.initialLayout = VK_IMAGE_LAYOUT_PREINITIALIZED;
+	img_info.extent = (VkExtent3D) { attribs->width, attribs->height, 1 };
+	img_info.usage = for_render ?
+		VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT :
+		VK_IMAGE_USAGE_SAMPLED_BIT;
+	if (disjoint) {
+		img_info.flags = VK_IMAGE_CREATE_DISJOINT_BIT;
+	}
+
+	VkExternalMemoryImageCreateInfo eimg = {0};
+	eimg.sType = VK_STRUCTURE_TYPE_EXTERNAL_MEMORY_IMAGE_CREATE_INFO;
+	eimg.handleTypes = VK_EXTERNAL_MEMORY_HANDLE_TYPE_DMA_BUF_BIT_EXT;
+	img_info.pNext = &eimg;
+
+	VkSubresourceLayout plane_layouts[WLR_DMABUF_MAX_PLANES] = {0};
+	VkImageDrmFormatModifierExplicitCreateInfoEXT mod_info = {0};
+
+	img_info.tiling = VK_IMAGE_TILING_DRM_FORMAT_MODIFIER_EXT;
+	for (unsigned i = 0u; i < plane_count; ++i) {
+		plane_layouts[i].offset = attribs->offset[i];
+		plane_layouts[i].rowPitch = attribs->stride[i];
+		plane_layouts[i].size = 0;
+	}
+
+	mod_info.sType = VK_STRUCTURE_TYPE_IMAGE_DRM_FORMAT_MODIFIER_EXPLICIT_CREATE_INFO_EXT;
+	mod_info.drmFormatModifierPlaneCount = plane_count;
+	mod_info.drmFormatModifier = mod->props.drmFormatModifier;
+	mod_info.pPlaneLayouts = plane_layouts;
+	eimg.pNext = &mod_info;
+
+	VkImage image;
+	res = vkCreateImage(dev, &img_info, NULL, &image);
+	if (res != VK_SUCCESS) {
+		wlr_vk_error("vkCreateImage", res);
+		return VK_NULL_HANDLE;
+	}
+
+	unsigned mem_count = disjoint ? plane_count : 1u;
+	VkBindImageMemoryInfo bindi[WLR_DMABUF_MAX_PLANES] = {0};
+	VkBindImagePlaneMemoryInfo planei[WLR_DMABUF_MAX_PLANES] = {0};
+
+	for (unsigned i = 0u; i < mem_count; ++i) {
+		struct VkMemoryFdPropertiesKHR fdp = {0};
+		fdp.sType = VK_STRUCTURE_TYPE_MEMORY_FD_PROPERTIES_KHR;
+		res = renderer->dev->api.getMemoryFdPropertiesKHR(dev, htype,
+			attribs->fd[i], &fdp);
+		if (res != VK_SUCCESS) {
+			wlr_vk_error("getMemoryFdPropertiesKHR", res);
+			goto error_image;
+		}
+
+		VkImageMemoryRequirementsInfo2 memri = {0};
+		memri.image = image;
+		memri.sType = VK_STRUCTURE_TYPE_IMAGE_MEMORY_REQUIREMENTS_INFO_2;
+
+		VkImagePlaneMemoryRequirementsInfo planeri = {0};
+		if (disjoint) {
+			planeri.sType = VK_STRUCTURE_TYPE_IMAGE_PLANE_MEMORY_REQUIREMENTS_INFO;
+			planeri.planeAspect = mem_plane_aspect(i);
+			memri.pNext = &planeri;
+		}
+
+		VkMemoryRequirements2 memr = {0};
+		memr.sType = VK_STRUCTURE_TYPE_MEMORY_REQUIREMENTS_2;
+
+		renderer->dev->api.getImageMemoryRequirements2(dev, &memri, &memr);
+		int mem = vulkan_find_mem_type(renderer->dev, 0,
+			memr.memoryRequirements.memoryTypeBits & fdp.memoryTypeBits);
+		if (mem < 0) {
+			wlr_log(WLR_ERROR, "no valid memory type index");
+			goto error_image;
+		}
+
+		// since importing transfers ownership of the fd to vulkan, we have
+		// to duplicate it since this operation does not transfer ownership
+		// of the attribs to this texture. Will be closed by vulkan on
+		// vkFreeMemory (i guess; could not find it in spec)
```

Comment:
That is indeed the case. You dup() the FD you get if you want to make a reference (ownership is explicitly transferred), and the driver is responsible for closing it when you free the memory.